### PR TITLE
Add support for system roles and site-admin on custom course nodes

### PR DESCRIPTION
**Pull request by adpe**
Hello @abias 

I just made a first draft. It does what it should, but would be happy about a code review. There are now two additional parameters to define the global roles and the site admin status.

Best regards
Adrian

### CODE REVIEW

**Comment by adpe** [comment]
Hello @abias
I want to ask when you plan to review this PR? We're not using it yet in production with these changes.
Would be nice if you could check it out.

File: locallib.php [hunk 1]
```diff
@@ -103,14 +103,15 @@ function local_boostnavigation_build_custom_nodes($customnodes, navigation_node
         $nodevisible = false;
         $nodeischild = false;
         $nodekey = null;
+        $nodelanguage = null;
```
Comment (abias) [inline locallib.php:106]: This change is fine as I must have forgotton this line when I added the language parameter, but it isn't related to this particular feature. Could you please split it into a separate commit with a commit message like "Improve processing of the language parameter"?

File: locallib.php [hunk 2]
```diff
@@ -149,7 +150,8 @@ function local_boostnavigation_build_custom_nodes($customnodes, navigation_node
                             // Only proceed if something is entered here. This parameter is optional.
                             // If no language is given the node will be added to the navigation by default.
                             $nodelanguages = array_map('trim', explode(',', $setting));
-                            $nodevisible &= in_array(current_language(), $nodelanguages);
+                            $nodelanguage = in_array(current_language(), $nodelanguages);
+                            $nodevisible &= $nodelanguage;
```
Comment (abias) [inline locallib.php:154]: Is there a reason that you did split this line, but effectively don't change the functionality?

File: locallib.php [hunk 3]
```diff
@@ -422,3 +446,42 @@ function local_boostnavigation_user_has_role_on_page($userid, $setting) {
...
+ * @param string $setting A value which describes the state, e.g. true.
+ * @return bool
+ */
+function local_boostnavigation_user_is_admin($userid, $setting) {
+    if ($setting == "true") {
+        return is_siteadmin($userid);
+    } else {
+        return false;
+    }
+}
```
Comment (abias) [inline locallib.php:487]: It would be nice if there would still be an empty line at the end of the file as it is best practice in Moodle development.

File: lang/en/local_boostnavigation.php [hunk 4]
```diff
@@ -117,6 +117,8 @@
...
 <li><b>Supported cohort(s) (optional):</b> This setting can be used for displaying the custom node to members of the specified cohort only. Use the cohort\'s ID, not the cohort\'s name, for this setting. Separate more than one supported cohort with commas. If the custom node should be displayed for users regardless of any cohort membership, then leave this field empty.</li>
 <li><b>Supported role(s) (optional):</b> This setting can be used for displaying the custom node only to members with the specified role in each context. Use the role\'s shortname for this setting. Separate more than one supported role with commas. If the custom node should be displayed for users regardless of any role, then leave this field empty.</li>
+<li><b>Supported global role(s) (optional):</b> This setting can be used for displaying the custom node only to users with the specified global role. Use the role\'s shortname for this setting. Separate more than one supported role with commas. If the custom node should not be displayed for users with global role(s), then leave this field empty.</li>
```
Comment (abias) [inline lang/en/local_boostnavigation.php:120]: Currently, the 5th parameter only processes roles which are assigned on the course level. So, this addition for system roles generally makes sense for some setups and I am happy to add it.

There are two things which I would like to mention:
* Please let's use the term "system roles" and not "global roles" as this is the term used in Moodle core on /admin/roles/assign.php?contextid=1
* Your sentence "If the custom node should not be displayed for users with global role(s), then leave this field empty." is wrong. The node is added to the navigation unless it is restricted to certain roles. Thus, you don't hide the node if this field is empty. Please rephrase it as "If the custom node should be displayed for users regardless of any system role, then leave this field empty." which is then the same as for the fifth parameter.

File: lang/en/local_boostnavigation.php [hunk 5]
```diff
@@ -117,6 +117,8 @@
...
+<li><b>Supported admin (optional):</b> This setting can be used for displaying the custom node additionally for site-admins. Use "true" for this setting. If the custom node should not be displayed for admin users, then leave this field empty.</li>
```
Comment (abias) [inline lang/en/local_boostnavigation.php:121]: I hesitate to add the admin role as a dedicated parameter. Basically, the admin role is just a special role which is assigned on the system level.

Don't you think that we could omit this parameter and add "admin" as a fixed role to the sixth parameter?

File: locallib.php [hunk 6]
```diff
@@ -167,6 +169,28 @@ function local_boostnavigation_build_custom_nodes($customnodes, navigation_node
                             // so that the custom node is displayed.
                             $nodevisible &= local_boostnavigation_user_has_role_on_page($USER->id, $setting);
 
+                            break;
+                        // Check for the optional sixth parameter: system role filter.
+                        case 5:
+                            // Only proceed if some role is entered here. This parameter is optional.
+                            // If no system role shortnames are given, the node will not be added to the navigation by default.
```
Comment (abias) [inline locallib.php:176]: Similar to my comment above, this comment seems to be wrong.
It should be:
"If no system role shortnames are given, the node will be added to the navigation by default."

File: locallib.php [hunk 7]
```diff
@@ -167,6 +169,28 @@ function local_boostnavigation_build_custom_nodes($customnodes, navigation_node
...
+                            // Otherwise, it is checked whether the user has any of the provided roles,
+                            // so that the custom node is displayed.
+                            if (local_boostnavigation_user_has_role_on_system($USER->id, $setting) && $nodelanguage) {
```
Comment (abias) [inline locallib.php:179]: Why do you include $nodelanguage here? This is counterproductive from my point of view.
Couldn't you just use the same syntax as used in line https://github.com/moodleuulm/moodle-local_boostnavigation/pull/17/files#diff-528c47baf0e31a7dcd6c3042a2182b9cR170 ?

File: locallib.php [hunk 8]
```diff
@@ -167,6 +169,28 @@ function local_boostnavigation_build_custom_nodes($customnodes, navigation_node
...
+                                $nodevisible = true;
+                            }
+
+                            break;
+                        // Check for the optional seventh parameter: admin filter.
```
Comment (abias) [inline locallib.php:184]: As written above: I would be happy to see this dedicated parameter included in the sixth parameter

File: locallib.php [hunk 9]
```diff
@@ -422,3 +446,42 @@ function local_boostnavigation_user_has_role_on_page($userid, $setting) {
...
+    return false;
+}
+
+/**
+ * Checks if the user is a site-admin and the parameter is set true.
+ * @param int $userid
+ * @param string $setting A value which describes the state, e.g. true.
+ * @return bool
+ */
+function local_boostnavigation_user_is_admin($userid, $setting) {
```
Comment (abias) [inline locallib.php:481]: As written above: I would be happy to see this dedicated parameter included in the sixth parameter

File: locallib.php [hunk 10]
```diff
@@ -422,3 +446,42 @@ function local_boostnavigation_user_has_role_on_page($userid, $setting) {
         return count(array_intersect($rolesincontextshortnames, $showforroles)) > 0;
     }
 }
+
+/**
+ * Checks if the user has any of the allowed global system roles.
+ * @param int $userid
+ * @param string $setting A comma-separated whitelist of allowed system roles.
```
Comment (abias) [inline locallib.php:453]: This should be "[...] allowed system role shortnames."

File: locallib.php [hunk 11]
```diff
@@ -422,3 +446,42 @@ function local_boostnavigation_user_has_role_on_page($userid, $setting) {
...
+ * @return bool
+ */
+function local_boostnavigation_user_has_role_on_system($userid, $setting) {
+
+    $hasrole = null;
```
Comment (abias) [inline locallib.php:458]: I think this line should be dropped as you don't use this variable anywhere.

**Review by abias — COMMENTED** [review]
Hi @adpe ,

I am sorry that I didn't have the time to look at your code earlier. Shame on me.

I have added several comments to the code and would be happy if you could process them.
If you have any questions, please let me know.

Thanks,
Alex

File: locallib.php [hunk 12]
```diff
@@ -422,3 +446,42 @@ function local_boostnavigation_user_has_role_on_page($userid, $setting) {
...
+
+    // Split optional setting by comma.
+    $showforroles = explode(',', $setting);
+
+    // Get roles for user.
+    $roles = get_user_roles(context_system::instance(), $userid, false);
```
Comment (abias) [inline locallib.php:464]: Why don't you want to use the cache by adding false as third parameter?

File: locallib.php (hunk 2 above)
Comment (abias) [reply locallib.php:154]: OK, I see. This line is connected with your change in line 106.

**Comment by abias** [comment]
Hi Adrian,

as I wanted to proceed with #14 which should land after this change here, I have finished it myself in https://github.com/moodleuulm/moodle-local_boostnavigation/commit/2844d3d377e2ec2511b33a3b8cfdf2f14cb8ed8c. I have added your name to the Changelog and to the commit message.

Thank you for your contribution!

Cheers,
Alex

**Comment by adpe** [comment]
Many thanks!
